src/shopper-insights/shopperSession.js

```diff
@@ -0,0 +1,340 @@
+|};
+
+type RecommendedPaymentMethodsResponse = {|
+  eligible_methods: {
+    [paymentMethod: "paypal" | "venmo"]: {|
+      can_be_vaulted: boolean,
+      eligible_in_paypal_network?: boolean,
+      recommended?: boolean,
+      recommended_priority?: number,
+    |},
+  },
+|};
+
+type SdkConfig = {|
+  sdkToken: ?string,
+  pageType: ?string,
+  userIDToken: ?string,
+  clientToken: ?string,
+  paypalApiDomain: string,
+  environment: ?string,
+  buyerCountry: string,
+  currency: string,
+|};
+
+// eslint's flow integration is very out of date
+// it doesn't recognize the generics here as used
+// eslint-disable-next-line no-undef
+type Request = <TRequestData, TResponse>({|
+  method?: string,
+  url: string,
+  // eslint-disable-next-line no-undef
+  data: TRequestData,
+  accessToken: ?string,
+  // eslint-disable-next-line no-undef
+|}) => Promise<TResponse>;
+
+type Storage = {|
+  // eslint's flow integration is very out of date
+  // it doesn't recognize the generics here as used
+  // eslint-disable-next-line no-undef
+  get: <TValue>(key: string) => ?TValue,
+  // eslint-disable-next-line flowtype/no-weak-types
+  set: (key: string, value: any) => void,
+|};
+
+const parseEmail = (merchantPayload): ?{| email: string |} => {
+  if (!merchantPayload.email) {
+    return;
+  }
+
+  const email = merchantPayload.email;
+  const isValidEmail =
+    typeof email === "string" && email.length < 320 && /^.+@.+$/.test(email);
+
+  if (!isValidEmail) {
+    throw new ValidationError(
+      `Expected shopper information to include a valid email format`
+    );
+  }
+
+  return {
+    email,
+  };
+};
+
+const parsePhone = (
+  merchantPayload
+): ?{| phone: {| country_code: string, national_number: string |} |} => {
+  if (!merchantPayload.phone) {
+    return;
+  }
+
+  if (
+    !merchantPayload.phone.nationalNumber ||
+    !merchantPayload.phone.countryCode
+  ) {
+    throw new ValidationError(
+      `Expected phone number for shopper insights to include nationalNumber and countryCode`
+    );
+  }
+
+  const nationalNumber = merchantPayload.phone.nationalNumber;
+  const countryCode = merchantPayload.phone.countryCode;
+  const isValidPhone =
+    typeof nationalNumber === "string" && /\d{5,}/.test(nationalNumber);
+
+  if (!isValidPhone) {
+    throw new ValidationError(
+      `Expected shopper information to be a valid phone number format`
+    );
+  }
+
+  return {
+    phone: {
+      country_code: countryCode,
+      national_number: nationalNumber,
+    },
+  };
+};
+
+export const parseMerchantPayload = ({
+  merchantPayload,
+  sdkConfig,
+}: {|
+  merchantPayload: MerchantPayloadData,
+  sdkConfig: SdkConfig,
+|}): RecommendedPaymentMethodsRequestData => {
+  const email = parseEmail(merchantPayload);
+  const phone = parsePhone(merchantPayload);
+
+  return {
+    customer: {
+      ...(sdkConfig.environment !== "production" && {
+        country_code: sdkConfig.buyerCountry,
+      }),
+      // $FlowIssue too many cases?
```

Comment:
this ^ happened to me recently - frustrating! (maybe we can ignore it in the near future)